### ISSUE TITLE
Azcopy uses excessive amount of memory

### ISSUE THREAD

**Issue by shchukax** [Azure/azure-storage-azcopy#2051]
### Which version of the AzCopy was used? 
10.17.0

### Which platform are you using? (ex: Windows, Mac, Linux)
Windows Server 2019 in Azure VM

### What command did you run?
```
azcopy copy "https://xxxxx.file.core.windows.net/media/*?sas-reducted" "https://xxxxx.blob.core.windows.net/container/?sas-reducted" --recursive --log-level WARNING --output-level quiet
```

### What problem was encountered?
The machine, on which azcopy is running has very limited memory - 4GB total, of which around 3 GB is used by other processes.  These are existing constraints.  I use environment variables to control certain aspects, including:
```
AZCOPY_CONCURRENCY_VALUE = 16
AZCOPY_BUFFER_GB=0.1
```
I expect azcopy to take no more than 0.5 GB, however it's using significantly more:
![image](https://user-images.githubusercontent.com/9897788/216596587-616f80d6-dc65-4a79-921e-53b676b59dd4.png)

azcopy then dies with "out of memory" error after copying about 50K files.

I have tried playing with the two environment variables above and used `azcopy env` to check that they are picked up (they are).  However I couldn't get azcopy to use less than 1.5GB no matter what I tried.  I need to get it to run in under 0.5GB total.

### How can we reproduce the problem in the simplest way?
File share being azcopied to blob container contains about 3-4 million small files in tens of thousands of directories, many levels deep.  Set the parameters as described and use the command above.  

### Have you found a mitigation/solution?
No

**Comment by nakulkar-msft** [comment]
Hi, AzCopy is resource intensive. Memory is used to parallelize transfers and improve transfer speed. Please provide  a better environment for the transfer.

**Comment by marc-hb** [comment]
Similar tools like rsync and many others can transfer millions of files quickly and in parallel while using a fraction of the memory. Please catch up with the "competition".

https://rsync.samba.org/FAQ.html#4

> memory usage

> Rsync versions before 3.0.0 always build the entire list of files to be transferred at the beginning and hold it in memory for the entire run. Rsync needs about 100 bytes to store all the relevant information for one file, so (for example) a run with 800,000 files would consume about 80M of memory. -H and --delete increase the memory usage further.

> Version 3.0.0 slightly reduced the memory used per file by not storing fields not needed for a particular file. It also introduced an incremental recursion mode that builds the file list in chunks and holds each chunk in memory only as long as it is needed. This mode dramatically reduces memory usage, but it only works provided that both sides are 3.0.0 or newer and certain options that rsync currently can't handle in this mode are not being used.
